lib/client/tncon/buffer_test.go

```diff
@@ -0,0 +1,182 @@
+}
+
+func TestBufferedChannelPipeWrite(t *testing.T) {
+	// With a sufficient buffer, write should successfully
+	// write to the channel without blocking
+	for _, tc := range []struct {
+		buffer int
+		len    int
+	}{
+		{
+			buffer: 0,
+			len:    0,
+		}, {
+			buffer: 0,
+			len:    10,
+		}, {
+			buffer: 10,
+			len:    10,
+		}, {
+			buffer: 10,
+			len:    100,
+		}, {
+			buffer: 100,
+			len:    100,
+		},
+	} {
+		t.Run(fmt.Sprintf("buffer=%v, len=%v", tc.buffer, tc.len), func(t *testing.T) {
+			buffer := newBufferedChannelPipe(tc.buffer)
+			t.Cleanup(func() { require.NoError(t, buffer.Close()) })
+
+			// drain channel
+			rc := make(chan []byte)
+			go func() {
+				read := make([]byte, tc.len)
+				for i := range read {
+					read[i] = <-buffer.ch
+				}
+				rc <- read
+			}()
+
+			p := make([]byte, tc.len)
+			for n := 0; n < tc.len; n++ {
+				p[n] = byte(n)
+			}
+
+			n, err := buffer.Write(p)
+			require.NoError(t, err)
+			require.Equal(t, tc.len, n)
+			require.Equal(t, p, <-rc)
+		})
+	}
+}
+
+func TestBufferedChannelPipeRead(t *testing.T) {
+	for _, tc := range []struct {
+		desc     string
+		buffer   int
+		writeLen int
+		readLen  int
+		expectN  int
+	}{
+		{
+			desc:     "empty read",
+			buffer:   0,
+			writeLen: 0,
+			readLen:  0,
+			expectN:  0,
+		}, {
+			desc:     "one byte read",
+			buffer:   0,
+			writeLen: 1,
+			readLen:  1,
+			expectN:  1,
+		}, {
+			desc:     "read with sufficient buffer",
+			buffer:   10,
+			writeLen: 10,
+			readLen:  10,
+			expectN:  10,
+		}, {
+			desc:     "large read with sufficient buffer",
+			buffer:   255,
+			writeLen: 255,
+			readLen:  255,
+			expectN:  255,
+		}, {
+			desc:     "read with smaller write",
+			buffer:   100,
+			writeLen: 10,
+			readLen:  100,
+			expectN:  10,
+		}, {
+			desc:     "read with larger write",
+			buffer:   100,
+			writeLen: 100,
+			readLen:  10,
+			expectN:  10,
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			buffer := newBufferedChannelPipe(tc.buffer)
+			t.Cleanup(func() { require.NoError(t, buffer.Close()) })
+
+			write := make([]byte, tc.writeLen)
+			for i := 0; i < tc.writeLen; i++ {
+				write[i] = byte(i)
+			}
+
+			// fill channel
+			go buffer.Write(write)
+
+			p := make([]byte, tc.readLen)
+			n, err := buffer.Read(p)
+			require.NoError(t, err)
+			require.Equal(t, tc.expectN, n)
+			require.Equal(t, write[:n], p[:n])
+		})
+	}
+}
+
+// BenchmarkBufferedChannelPipe is a benchmark test for writing
+// to a buffer, reading from it, and closing it.
+func BenchmarkBufferedChannelPipe(b *testing.B) {
+	for n := 0; n < b.N; n++ {
+		b.Run(fmt.Sprint(n), func(b *testing.B) {
+			buffer := newBufferedChannelPipe(sequenceBufferSize)
+
+			errCh := make(chan error)
+			go func() {
+				_, err := io.ReadAll(buffer)
+				errCh <- err
+			}()
+
+			b.Cleanup(func() {
+				require.NoError(b, buffer.Close())
+				require.NoError(b, <-errCh)
+			})
+			buffer.Write(make([]byte, 100*sequenceBufferSize))
+		})
+	}
```

Comment:
Looks good, thanks! I adapted it to include `buffer.Read` in the benchmark, since this is actually the benchmark we care about (reading terminal input).

Reply:
I also updated it to always write 1000 bytes of data, to make it show a better comparison:
```
BenchmarkBufferedChannelPipe/size=1-12         	    4215	    330877 ns/op	    1016 B/op	       2 allocs/op
BenchmarkBufferedChannelPipe/size=10-12        	    7221	    244740 ns/op	    1016 B/op	       2 allocs/op
BenchmarkBufferedChannelPipe/size=100-12       	    8412	    143209 ns/op	    1016 B/op	       2 allocs/op
BenchmarkBufferedChannelPipe/size=500-12       	    7804	    133376 ns/op	    1016 B/op	       2 allocs/op
BenchmarkBufferedChannelPipe/size=1000-12      	    8066	    134231 ns/op	      16 B/op	       2 allocs/op
```